Hot Rod map storage: User / client session no-downtime store

martin-kanis opened the pull request:
I introduced `HotRodStringPair` where `key` and `value` are both type `String` instead of `WrappedMessage`. Reason is that we are not able to use `WrappedMessage` in Ickle Query. See following error:
`ISPN028504 The property authenticatedClientSessions.key is an embedded entity and does not allow comparison predicates`

Question:
Should we use indexes for HotRodStringPair and where we don't use HotRodPair in a query we will use HotRodPair<String, String> instead?

Closes #9675

martin-kanis committed:
Hot Rod map storage: User / client session no-downtime store

abstractj commented:
@martin-kanis CodeQL is workflow due to build failures. Do you have any idea about the errors below?

```
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  05:07 min
[INFO] Finished at: 2022-04-08T09:11:19Z
[INFO] ------------------------------------------------------------------------
Error:  Failed to execute goal on project keycloak-model-map-hot-rod: Could not resolve dependencies for project org.keycloak:keycloak-model-map-hot-rod:jar:999-SNAPSHOT: Could not find artifact org.keycloak:keycloak-model-map:jar:tests:999-SNAPSHOT -> [Help 1]
Error:  
Error:  To see the full stack trace of the errors, re-run Maven with the -e switch.
Error:  Re-run Maven using the -X switch to enable full debug logging.
Error:  
Error:  For more information about the errors and possible solutions, please read the following articles:
Error:  [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/DependencyResolutionException
Error:  
Error:  After correcting the problems, you can resume the build with the command
Error:    mvn <args> -rf :keycloak-model-map-hot-rod
Error: Process completed with exit code 1.
```

martin-kanis commented:
@abstractj It's caused by the fact we don't build test related artifacts in this workflow (see build command `mvn install -Dmaven.test.skip -DskipQuarkus -DskipTestsuite -DskipExamples -DskipTests`) and there is a test dependency https://github.com/keycloak/keycloak/blob/main/model/map-hot-rod/pom.xml#L25-L26 in `keycloak-model-map-hot-rod`. 
I think it can be fixed by removing `-Dmaven.test.skip` from the build command but I'm not sure if that doesn't have some unwanted side effects. 
CC @mhajas

abstractj commented:
@martin-kanis @mhajas fixed it here https://github.com/keycloak/keycloak/pull/11193

mhajas commented on model/map-hot-rod/src/main/java/org/keycloak/models/map/storage/hotRod/userSession/HotRodSessionState.java:
```diff
@@ -0,0 +1,46 @@
+        LOGGED_OUT_UNCONFIRMED
+    }
+
+    enum PersistenceState {
```
I am wondering, do we need this field here? If we store an entity it is always PERSISTENT, isn't it?

mhajas replied:
I see. But what if we have something like this in the `AbstractHotRodUserSessionEntityDelegate`:
```
        @Override
        public UserSessionModel.SessionPersistenceState getPersistenceState() {
            return UserSessionModel.SessionPersistenceState.PERSISTENT;
        }

        @Override
        public void setPersistenceState(UserSessionModel.SessionPersistenceState persistenceState) {
            if (UserSessionModel.SessionPersistenceState.PERSISTENT != persistenceState) {
                throw new IllegalArgumentException("Transient session should not be stored in the HotRod.");
            }
        }
```

If I understand it correctly, it would still work and we would save some space in the Infinispan server. I am not sure it is correct, WDYT?

martin-kanis replied:
This could work, but currently it fails with `Cannot find HotRodUserSessionEntity field for methods: [getPersistenceState, setPersistenceState]` if I don't have some field called `persistenceState` in `HotRodUserSessionEntity`.

mhajas replied:
Hmm, I see. It is because of this: https://github.com/keycloak/keycloak/blob/main/model/build-processor/src/main/java/org/keycloak/models/map/processor/GenerateHotRodEntityImplementationsProcessor.java#L272-L283 Maybe we can move it somewhere else, or just remove it. The build would fail anyway if some field that doesn't exist is used.

martin-kanis replied:
Thanks for the suggestion. I removed only the `if` check because `hotRodVariable` is used later, but not sure we can leave without a `null` check. Also I added `null` check here: `if (UserSessionModel.SessionPersistenceState.PERSISTENT != persistenceState)`

mhajas commented on model/map-hot-rod/src/main/java/org/keycloak/models/map/storage/hotRod/userSession/HotRodUserSessionEntity.java:
```diff
@@ -0,0 +1,152 @@
+        }
+
+        @Override
+        public UserSessionModel.State getState() {
```
It should be possible to implement these methods by the generator. Previously, it was not possible because we stored these in the `Map<String, Enum>`, however, it should work with a simple field.

martin-kanis replied:
done

mhajas reviewed (changes requested):
Thank you for the PR. 

Regarding `HotRodStringPair`. The reason for introducing the class is to add the possibility for searching by `String-String` pairs. Since all searched fields should have an index on them, we should add indexes there and we should use the class only in places where we need to search by the field, otherwise we can use `HotRodPair<String, String>

mhajas commented:
@martin-kanis Could you please have a look at this comment:
> Regarding HotRodStringPair. The reason for introducing the class is to add the possibility for searching by String-String pairs. Since all searched fields should have an index on them, we should add indexes there and we should use the class only in places where we need to search by the field, otherwise we can use `HotRodPair<String, String>`

mhajas reviewed (changes requested):
Sorry for missing that @martin-kanis.

Please add indexes also for following fields that are used in criteria builder:
`HotRodUserSessionEntity.lastSessionRefresh`
`HotRodUserSessionEntity.offline`
`HotRodAuthenticatedClientSessionEntity.timestamp`
`HotRodAuthenticatedClientSessionEntity.offline`

martin-kanis commented:
> @martin-kanis Could you please have a look at this comment:
> 
> > Regarding HotRodStringPair. The reason for introducing the class is to add the possibility for searching by String-String pairs. Since all searched fields should have an index on them, we should add indexes there and we should use the class only in places where we need to search by the field, otherwise we can use `HotRodPair<String, String>`

Sorry I missed this comment. It's now addressed.

> Please add indexes also for following fields that are used in criteria builder:
HotRodUserSessionEntity.lastSessionRefresh
HotRodUserSessionEntity.offline
HotRodAuthenticatedClientSessionEntity.timestamp
HotRodAuthenticatedClientSessionEntity.offline

Added, except on the `HotRodAuthenticatedClientSessionEntity.timestamp` as it is currently not used in any query.

martin-kanis commented:
Model test failure is not related to the PR.

abstractj commented:
@martin-kanis the Quickstarts failures are unrelated with your change. We have #10950 to address this issue. About the model tests, not sure if they are related or not.

mhajas reviewed (approved):
Thanks

hmlnarik reviewed (approved):
Approving per @mhajas 's review. Test failures are unrelated
